--- head/sys/netpfil/ipfw/ip_fw_pfil.c	Wed Apr 16 12:29:45 2014	(r264539)
+++ head/sys/netpfil/ipfw/ip_fw_pfil.c	Wed Apr 16 14:37:11 2014	(r264540)
@@ -334,7 +334,7 @@ ipfw_check_frame(void *arg, struct mbuf 
m_adj(m, ETHER_HDR_LEN);	/* strip ethernet header */

args.m = m;	 /* the packet we are looking at	 */
-	args.oif = dst;	 /* destination, if any	 */
+	args.oif = dir == PFIL_OUT ? dst: NULL;	/* destination, if any	*/
args.next_hop = NULL;	/* we do not support forward yet	*/
args.next_hop6 = NULL;	/* we do not support forward yet	*/
args.eh = &save_eh;	/* MAC header for bridged/MAC packets	*/